### PR TITLE
Associations

### PR DESCRIPTION
**Pull request by NataJenkins**
Rails asociations to:

Create an event with an user, and be able to invite other users to the events.
users must be logged in to accept the invite to the events.

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 2021_01_05_121953) do
+
+  create_table "attendees", force: :cascade do |t|
+    t.string "user_id"
+    t.string "event_id"
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+  end
```
Comment (ghost) [inline db/schema.rb]: Index the `user_id` and `event_id` keys.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required ✖️ 

### Hello team 😄,

**Great job on your project 💪. However, there a few fixes needed.**

1). **When I run `rails db:migrate`, I get the error shown below.**
migration

2). In the user-events associations, the user should be referenced as "creator" in the events model. Also, the events table should have the foreign key `creator_id` instead of `user_id`. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from Odin.

3).**The join model, `attendee.rb`, is incomplete. It should have something like the following**

```
class Attendee < ApplicationRecord
  belongs_to :attendee, class_name: "User"
  belongs_to :attended_event, class_name: "Event"
end
```
see [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/attendee_assoc_requirement.png)t from Odin. 

4). **Since you are using a join model, attendee.rb, for your many-to-many relationships, you have to use the `has many through` method in user.rb and events.rb, to declare the associations. For example, in your user.rb model, you should have another line like the following**

```
class User < ActiveRecord::Base
has_many :attendees
has_many :attended_events, through: :attendees
end
```
You can reference this [article](https://guides.rubyonrails.org/association_basics.html#choosing-between-has-many-through-and-has-and-belongs-to-many) on how to use `has_many :through`. Please remember it has to be used in `events.rb` file too.

5). **The events index page should list all events separated into past and upcoming categories. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from Odin.**

6). **The user show page should list events created by user(see [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from Odin), previously attended by user and upcoming events to be attended by user. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from Odin.**

7).**There should be no conditional logic(if statements and each loops) in your views directory files. Please remove it and instead write it in Views helpers `(app/helpers/*_helper.rb)`. You can reference the following articles on this: [article1](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers), [article2](http://codefol.io/posts/Where-Do-I-Put-My-Code/).**

8). **Add Rspec tests as described in the screenshot below from pathwright.**
rspec_tests

9). **Please see one comment below.**


### Optional
1). Define routes for only used actions.

## Happy coding 🥂 

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by jubaan — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [x] When running the `rails server` command, this error message appears on the browser
image
- [x] Trying to figure out the cause and commenting the past line (not the best solution) of code also throws an error with line `9`  of the `./app/views/layouts/application.html.erb`
image
  - NOTE: Commenting both lines let me run the application on the browser. Please find out what's going on and make it work with the code in the repository or change if necessary.

- [x] Users can't logout
image

- [x] 3).The join model, `attendee.rb`, is incomplete. It should have something like the following **STILL PENDING**
> 
> ```
> class Attendee < ApplicationRecord
>   belongs_to :attendee, class_name: "User"
>   belongs_to :attended_event, class_name: "Event"
> end
> ```

> see [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/attendee_assoc_requirement.png)t from Odin.
>

- [x] 5). The events index page should list all events separated into past and upcoming categories. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from Odin.

  **STILL PENDING:  I've seen that when creating an event it identifies each new event as `past` or `future` event, this is great but still is not separating past from upcoming, as a recommendation create on the view a list for each and depending on the date will the event will be added to this list.**

- [x] 6). The user show page should list events created by user(see [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from Odin), previously attended by user and upcoming events to be attended by user. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from Odin.

  **STILL PENDING:  Please fix first the logout issue and check this point is working before submitting for a re-review.** 

- [ ] 7).There should be no conditional logic(if statements and each loops) in your views directory files. Please remove it and instead write it in Views helpers `(app/helpers/*_helper.rb)`. You can reference the following articles on this: [article1](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers), [article2](http://codefol.io/posts

  **STILL PENDING: To solve this create helpers that will accept an argument and will apply the logic in the same way you have it the views, the difference is that the helper will be stored in the `helpers` section of the application structure. A helper is just a ruby method that returns whatever you want. In this way instead of using logic inside the view, you will only call the newly created helper( or ruby method)**

- [ ] 8). Add Rspec tests as described in the screenshot below from pathwright.
> rspec_tests

  **STILL PENDING: Log out should also be tested and also elements related to creating, updating, viewing, and deleting events.**

### [OPTIONAL] :large_blue_circle: 
- [x] The Gem `wdm` could cause your project to stop working correctly on Linux or Unix OS please add this to your Gemfile so it doesn't crash when running `bundle install`
```
gem 'wdm', '>= 0.1.0' if Gem.win_platform?
```


## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

**Review by jubaan — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [x] When trying to create an attendee this error appears in the console
image
- [x] Event's index page displays past and upcoming **events lists**, as I mentioned before not just identify which one is _future_ and which one is _past_. They should be separated on a list.
  > * [x]  5). The events index page should list all events separated into past and upcoming categories. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from Odin.
  >   **STILL PENDING:  I've seen that when creating an event it identifies each new event as `past` or `future` event, this is great but still is not separating past from upcoming, as a recommendation create on the view a list for each and depending on the date will the event will be added to this list.**
- [x] There is no navigation with buttons to move around the page to move around the page
- [x] Event's show page doesn't display the list of attendees. As users can't invite other users to the event.
- [x] The user's show page displays events the user has created and events that the user is attending, also these events should be separated in past and upcoming
  > * 6). The user show page should list events created by user(see [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from Odin), previously attended by user and upcoming events to be attended by user. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from Odin.
  >   **STILL PENDING:  Please fix first the logout issue and check this point is working before submitting for a re-review.**
- [x] There should be no logic in the view, please review your views even the partials, and remove the logic from within.
  > * [x]  7).There should be no conditional logic(if statements and each loops) in your views directory files. Please remove it and instead write it in Views helpers `(app/helpers/*_helper.rb)`. You can reference the following articles on this: [article1](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers), [article2](http://codefol.io/posts
  >   **STILL PENDING: To solve this create helpers that will accept an argument and will apply the logic in the same way you have it the views, the difference is that the helper will be stored in the `helpers` section of the application structure. A helper is just a ruby method that returns whatever you want. In this way instead of using logic inside the view, you will only call the newly created helper( or ruby method)**

- [x] Integration testing is missing some specs
  > * 8). Add Rspec tests as described in the screenshot below from pathwright.
  > 
  > > rspec_tests
  > 
  > **STILL PENDING: Log out should also be tested and also elements related to creating, updating, viewing, and deleting events.**

## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

## Required Changes ♻️ 

- [x] after creating an event, I get an error whenever I try to visit the event's page, stopping me from completing my review
image
**_you're comparing a string to an integer, convert the string then compare_**

- [x] when I go to a specific event and click "create attendee" the button wouldn't work.
- [x] please remove the logic from your views ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
- [x] add more tests
  * [ ] models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
  * [ ] authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
  * [ ] event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Comment by mimipeshy** [comment]
### STATUS : CHANGES REQUIRED :red_circle: :recycle: 

Hey team @NataJenkins and @Nexch :wave: :wave: good job so far. However, a few things need to be completed before we can mark the project as complete 

- [x]  Event's index page should display past and upcoming events lists ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] The event's show page displays the list of attendants ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ]The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ] Navigation links are present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x]  when I go to a specific event and click "create attendee" the button wouldn't work.

That's all :smile: , kindly submit another review when done.
If you have any questions kindly drop me a message on slack i will be glad to help

Regards
Peris Ndanu

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work, team. 💻 👀 ✨ 

- Your app seems to work fine.

## However, there are some changes required 👀 🧰 🔨 🔧 

- I'm able to create events with blank dates. This in turn breaks the `/events` view because the application can't really display events with blank dates. Please fix this!

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by jubaan — APPROVED** [review]
# :tada: :tada: :tada: :rotating_light: Status: Approved :rotating_light:  :tada: :tada: :tada:

#### :clap: :clap: Great job :hammer_and_pick:!! There are no more issues left to be addressed. 

## See you at the top :rocket:!

# :tada: :tada: :tada: :rotating_light: Status: Approved :rotating_light:  :tada: :tada: :tada:

**Pull request merged**